File: backend/services/risk_engine.py
import numpy as np

class DynamicRiskEngine:
    """
    Real Dynamic Risk Score Engine.
    Computes mathematical normalized risk score (0-100) based on real acoustic signals & contextual parameters.
    """

    @staticmethod
    def calculate_risk(
        synthetic_probability: float,
        speaker_consistency: float,
        liveness_score: float,
        call_context_risk: float = 15.0,
        transaction_risk: float = 10.0,
        weights: dict = None
    ) -> dict:
        if weights is None:
            weights = {
                "synthetic_voice": 40.0,
                "speaker_mismatch": 25.0,
                "liveness_risk": 15.0,
                "call_context": 10.0,
                "transaction_risk": 10.0,
            }

        total_weight = sum(weights.values()) or 100.0

        w_synth = weights.get("synthetic_voice", 40.0) / total_weight
        w_speaker = weights.get("speaker_mismatch", 25.0) / total_weight
        w_liveness = weights.get("liveness_risk", 15.0) / total_weight
        w_context = weights.get("call_context", 10.0) / total_weight
        w_tx = weights.get("transaction_risk", 10.0) / total_weight

        speaker_mismatch_risk = max(0.0, 100.0 - speaker_consistency)
        liveness_deficit_risk = max(0.0, 100.0 - liveness_score)

        pt_synth = round(synthetic_probability * w_synth, 2)
        pt_speaker = round(speaker_mismatch_risk * w_speaker, 2)
        pt_liveness = round(liveness_deficit_risk * w_liveness, 2)
        pt_context = round(call_context_risk * w_context, 2)
        pt_tx = round(transaction_risk * w_tx, 2)

        raw_score = pt_synth + pt_speaker + pt_liveness + pt_context + pt_tx
        final_score = int(np.clip(round(raw_score), 0, 100))

        if final_score >= 81:
            level = "CRITICAL"
            recommended_action = "Block Sensitive Action & Place Transaction Hold Immediately"
        elif final_score >= 61:
            level = "HIGH"
            recommended_action = "Require Step-Up Verification (MFA / Trusted Callback)"
        elif final_score >= 31:
            level = "MEDIUM"
            recommended_action = "Display Operator Warning & Prompt Verification"
        else:
            level = "LOW"
            recommended_action = "Continue call under automated monitoring"

        primary_drivers = []
        if synthetic_probability >= 65:
            primary_drivers.append(f"High synthetic voice probability ({synthetic_probability}%)")
        if speaker_consistency <= 65:
            primary_drivers.append(f"Speaker consistency mismatch ({speaker_consistency}% similarity)")
        if liveness_score <= 70:
            primary_drivers.append(f"Low voice liveness score ({liveness_score}%)")
        if transaction_risk >= 60:
            primary_drivers.append(f"Sensitive transaction request ({transaction_risk}%)")
        if call_context_risk >= 60:
            primary_drivers.append(f"Unverified caller ID / SIP origin risk ({call_context_risk}%)")

        if not primary_drivers:
            primary_drivers.append("Interaction parameters within safe baseline")

        return {
            "final_score": final_score,
            "level": level,
            "weighted_breakdown": {
                "synthetic_pts": pt_synth,
                "speaker_pts": pt_speaker,
                "liveness_pts": pt_liveness,
                "context_pts": pt_context,
                "transaction_pts": pt_tx,
            },
            "recommended_action": recommended_action,
            "primary_drivers": primary_drivers,
        }

File: src/services/riskEngine.ts
import { RiskLevel, RiskScoreBreakdown, RiskWeights, VoiceAnalysisSignals } from '../types';
import { DEFAULT_RISK_WEIGHTS, RISK_BANDS } from '../utils/constants';

export class RiskEngine {
  public static calculateRisk(
    signals: VoiceAnalysisSignals,
    weights: RiskWeights = DEFAULT_RISK_WEIGHTS
  ): RiskScoreBreakdown {
    // Standardize total weights to ensure 100% normalized computation
    const totalWeight =
      weights.syntheticVoice +
      weights.speakerMismatch +
      weights.livenessRisk +
      weights.callContext +
      weights.transactionRisk;

    const normSyntheticW = (weights.syntheticVoice / (totalWeight || 100));
    const normSpeakerW = (weights.speakerMismatch / (totalWeight || 100));
    const normLivenessW = (weights.livenessRisk / (totalWeight || 100));
    const normContextW = (weights.callContext / (totalWeight || 100));
    const normTxW = (weights.transactionRisk / (totalWeight || 100));

    // Convert consistency and liveness to risk vectors
    const speakerMismatchRisk = Math.max(0, 100 - signals.speakerConsistency);
    const livenessRisk = Math.max(0, 100 - signals.livenessScore);

    // Compute weighted risk points
    const weightedSynthetic = Math.round(signals.syntheticProbability * normSyntheticW * 100) / 100;
    const weightedSpeaker = Math.round(speakerMismatchRisk * normSpeakerW * 100) / 100;
    const weightedLiveness = Math.round(livenessRisk * normLivenessW * 100) / 100;
    const weightedContext = Math.round(signals.callContextRisk * normContextW * 100) / 100;
    const weightedTransaction = Math.round(signals.transactionRisk * normTxW * 100) / 100;

    const rawScore = weightedSynthetic + weightedSpeaker + weightedLiveness + weightedContext + weightedTransaction;
    const finalScore = Math.min(100, Math.max(0, Math.round(rawScore)));

    let level: RiskLevel = 'LOW';
    let recommendedAction = RISK_BANDS.LOW.action;

    if (finalScore >= RISK_BANDS.CRITICAL.min) {
      level = 'CRITICAL';
      recommendedAction = RISK_BANDS.CRITICAL.action;
    } else if (finalScore >= RISK_BANDS.HIGH.min) {
      level = 'HIGH';
      recommendedAction = RISK_BANDS.HIGH.action;
    } else if (finalScore >= RISK_BANDS.MEDIUM.min) {
      level = 'MEDIUM';
      recommendedAction = RISK_BANDS.MEDIUM.action;
    }

    const primaryDrivers: string[] = [];

    if (signals.syntheticProbability >= 65) {
      primaryDrivers.push(`High synthetic voice probability (${signals.syntheticProbability}%)`);
    }
    if (signals.speakerConsistency <= 65) {
      primaryDrivers.push(`Speaker consistency mismatch (Only ${signals.speakerConsistency}% similarity)`);
    }
    if (signals.livenessScore <= 70) {
      primaryDrivers.push(`Low voice liveness score (${signals.livenessScore}%)`);
    }
    if (signals.transactionRisk >= 60) {
      primaryDrivers.push(`High risk transaction request detected (${signals.transactionRisk}%)`);
    }
    if (signals.callContextRisk >= 60) {
      primaryDrivers.push(`Unusual call origin / unverified Caller ID (${signals.callContextRisk}%)`);
    }

    if (primaryDrivers.length === 0) {
      primaryDrivers.push('Interaction parameters within safe baseline');
    }

    return {
      finalScore,
      level,
      weightedSynthetic,
      weightedSpeaker,
      weightedLiveness,
      weightedContext,
      weightedTransaction,
      recommendedAction,
      primaryDrivers,
    };
  }
}
